### PR TITLE
Add a new FFSK demodulator

### PR DESCRIPTION
**Pull request by radonyl**
China railway CIR (Cab Intergrated Radio) system use a special ffsk modulation scheme which uses 1200/1800Hz as carrier frequency, BCH(26, 16) as FEC algorithm and CRC16 as checksum. This implemtation is based on fmsfsk demodulator and new cirfsk finite state machine. Due to limited public document of CIR, the demodulator will directly output the raw hex content in CIR packages.

### CODE REVIEW

File: BCH26.c [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        {32, 32},
+        {16, 16},
+        {8, 8},
+        {4, 4},
+        {2, 2},
+        {1, 1}
+};
+
+unsigned int decode_BCH_26_16(unsigned int code, unsigned int *value){
+    // this code is from: https://blog.csdn.net/u012750235/article/details/84622161
```
Comment (weinholt) [inline BCH26.c]: I wonder, is this code under a GPL-compatible license? The custom for free software projects is that there should be a copyright notice in each file showing who holds the copyright, accompanied by license information. It is good to be careful about these things so it does not cause a problem later on.

File: BCH26.c [hunk 2]
```diff
@@ -1,6 +1,26 @@
...
+ *
+ *	This program is distributed in the hope that it will be useful,
+ *	but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *	GNU General Public License for more details.
+ *
+ *	You should have received a copy of the GNU General Public License
+ *	along with this program; if not, write to the Free Software
+ *	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
```
Comment (Arnie97) [inline BCH26.c]: I'm not a lawyer, but I'm sure that attaching the GPL copyright notices to a piece of proprietary code does not make it automagically GPL-compatible. Permission is required from the original author to a) make an adaptation, and b) does not given attribution to her or him in the copyright notice of the adoption work.

Maybe you can ask the author to share the code under a proper license. Or, you might just adapt a GPL-compatible alternative, say, https://github.com/torvalds/linux/blob/master/lib/bch.c.

Comment (Arnie97) [inline BCH26.c]: I'm not a lawyer, but I'm sure that attaching the GPL copyright notices to a piece of proprietary code does not make it automagically GPL-compatible. Permission is required from the original author to a) make an adaptation, and b) does not given attribution to her or him in the copyright notice of the adoption work.

Maybe you can ask the author to share the code under a proper license. Or, you might just adapt a GPL-compatible alternative, say, https://github.com/torvalds/linux/blob/master/lib/bch.c.

File: BCH26.c [hunk 3]
```diff
@@ -1,6 +1,26 @@
-#include <stdlib.h>
+/*
+ *      bch26.c -- BCH(26,16) Decoder
+ *      author: Alex L Manstein (alex.l.manstein@gmail.com)
+ *      Copyright (C) 2020
+ *
+ *	This program is free software; you can redistribute it and/or modify
+ *	it under the terms of the GNU General Public License as published by
+ *	the Free Software Foundation; either version 2 of the License, or
+ *	(at your option) any later version.
```
Comment (Arnie97) [inline BCH26.c]: I'm not a lawyer, but I'm sure that attaching the GPL copyright notices to a piece of proprietary code created by somebody else does not make it automagically GPL-compatible. Permission is required from the original author if you'd like to a) make an adaptation, and b) do not give attribution to her or him in the copyright notice of the adoption work.

Maybe you can ask the author to share the code under a proper license. Or, you might just adapt a GPL-compatible alternative, say, https://github.com/torvalds/linux/blob/master/lib/bch.c.

**Comment by EliasOenal** [comment]
I second what @Arnie97 said. There's actually a version of the Linux BCH code that was turned into am independent library, this could be worth a look: https://github.com/Parrot-Developers/bch

**Pull request closed without merging**